### PR TITLE
Add invoice page

### PR DESCRIPTION
**Pull request by albertkol**
## Done

- Add invoice endpoints `/account/invoices`
- Add invoice table
- Add invoice table filter
- Add `?email=example@email.com` to view as invoices

## QA

- Login at https://ubuntu-com-10062.demos.haus/advantage?test_backend=true
- Go to https://ubuntu-com-10062.demos.haus/account/invoices?test_backend=true
- Play around on the page: 
  - Check that you have the invoices
  - Test the filter
- Try to view as https://ubuntu-com-10062.demos.haus?test_backend=true&email=example@email.com

## Issue / Card

Fixes #52 #51 #58

### CODE REVIEW

**Comment by anthonydillon** [comment]
Could you update the fixes to the correct issues please.

File: templates/account/invoices/index.html
Comment (anthonydillon) [inline templates/account/invoices/index.html:18]: Question for UX. But wouldn't "Ubuntu Advantage" be better here inside of Canonical UA?

File: templates/account/invoices/index.html
Comment (anthonydillon) [inline templates/account/invoices/index.html:32]: Before looping could you check for length > 0. If not can you add an empty state row saying something like no invoices available.

File: webapp/advantage/schemas.py
Comment (anthonydillon) [inline webapp/advantage/schemas.py:76]: Should you have canonical-blender here too?

File: webapp/advantage/views.py
Comment (anthonydillon) [inline webapp/advantage/views.py:42]: From a branding point of view it's clearly Canonical and if not it's confusingly Canonical on Ubuntu website. Should this just be Ubuntu Advantage?

File: webapp/advantage/views.py
Comment (anthonydillon) [inline webapp/advantage/views.py:859]: Where does services come from? I don't see it in the function definition

**Review by anthonydillon — CHANGES_REQUESTED** [review]
Few comments in the code. You covered a lot in this PR. Could have been separate PRs.

**Comment by anthonydillon** [comment]
I just get a 500 page on step 2 of the QA steps.

**Comment by anthonydillon** [comment]
Do you not want UX or design review of this?

File: templates/account/invoices/index.html
Comment (albertkol) [reply templates/account/invoices/index.html:4]: I copy-pasted exactly what the `/payment-methods` page had in there. Which I assume is the same as what `/advantage` has. 

Perhaps we should look into creating a separate meta descriptions for each of the pages. Not sure who to ask for it. @lechsandecki @SirSamTumless 

But I'd like to not have the meta description block this PR. We can always update it as a new PR.

File: templates/account/invoices/index.html
Comment (albertkol) [reply templates/account/invoices/index.html:6]: Added.

File: templates/account/invoices/index.html
Comment (albertkol) [reply templates/account/invoices/index.html:18]: @wgx @lechsandecki thoughts?

We call the name of the service inside the table also `Canonical UA`.
image

File: webapp/advantage/schemas.py
Comment (albertkol) [reply webapp/advantage/schemas.py:76]: The API does not support it yet.

File: webapp/advantage/views.py
Comment (albertkol) [reply webapp/advantage/views.py:859]: I extracted it as a global variable at the beginning of `views.py`.
 
 https://github.com/canonical-web-and-design/ubuntu.com/pull/10062/files/60a15b39f47be74145a91c65efd7299d6d1d04ee#diff-aa9ecfa0f8c508e3003b04b0ffe83ba3698ed5f0a70237517932822a9a1a09e9R39-R52
 
I thought I should have a global place to define all the marketplaces to short name, full name conversions so they are consistent. I am not sure it's the best/nicest way to do it though.

File: templates/account/invoices/index.html
Comment (albertkol) [reply templates/account/invoices/index.html:32]: @wgx What would you like to show a user with no invoices? This is what I have at the moment.

image

**Comment by albertkol** [comment]
@anthonydillon I did not plan to solve 3 PRs in this one. The endpoints were 85% of the work. I added the table view to make the endpoints easier to QA. The filter was a quick thing to add on top.

Since the PR is not that big anyway, I found no reason to split it. But if you think it's necessary I can try to split it.

File: templates/account/invoices/index.html
Comment (wgx) [reply templates/account/invoices/index.html:32]: @albertkol this is fine

We also need a view / error (in the forthcoming acct mgmt story) for 

`You cannot view this page, please contact your account administrator for access.`

But not for now...

File: templates/account/invoices/index.html
Comment (wgx) [reply templates/account/invoices/index.html:18]: @albertkol this is fine for now, but @lechsandecki may want to change these

File: webapp/advantage/views.py
Comment (nottrobin) [reply webapp/advantage/views.py:859]: This is fair enough, but in this case you would normally want to signify that it's [a constant](https://www.python.org/dev/peps/pep-0008/#constants), with capital letters.

`SERVICES[payment_marketplace]["short"]`

To make it clear that it shouldn't be manipulated within any functions.

**Comment by anasereijo** [comment]
@albertkol this is looking great, just a couple of things:

- [ ] can you please add a bit of space between the title and the dropdown? (as per [design](https://app.zeplin.io/project/60be183358945711ff58fedc/screen/60bf27418133851582553084)) Maybe if you have 2 separate sections for the hero and content will work?
- [ ] this is my mistake but numbers on tables should be right-aligned as per our [guidelines](https://discourse.ubuntu.com/t/table-cell-contents/19858) - I updated the [design](https://zpl.io/29GwEXd)

Also, please make sure you always request a design review for new pages, new features or visual changes to pages/sections.
Thank you!

File: webapp/advantage/views.py
Comment (albertkol) [reply webapp/advantage/views.py]: Fixed!

**Comment by anasereijo** [comment]
looks good 🥇

**Comment by anthonydillon** [comment]
> @anthonydillon I did not plan to solve 3 PRs in this one. The endpoints were 85% of the work. I added the table view to make the endpoints easier to QA. The filter was a quick thing to add on top.
> 
> Since the PR is not that big anyway, I found no reason to split it. But if you think it's necessary I can try to split it.

No don't know but as you can see even resending a table can lead to back and forth so would be worth isolating in the future.

**Review by anthonydillon — CHANGES_REQUESTED** [review]
QA:
- I am getting a console error: `Uncaught TypeError: Currency code is required with currency style.` on the invoice page
- The invoices themselves are 500'ing for me in the demo (I remember you saying something about that this morning)

File: templates/account/invoices/index.html
Comment (anthonydillon) [reply templates/account/invoices/index.html:4]: As its no robots lets just make it not wrong. Something like "Your Canonical invoices for service you have signed up for."

File: templates/account/invoices/index.html
Comment (anthonydillon) [inline templates/account/invoices/index.html:3]: Also this would be better as something like "Your Canonical invoices"

Comment (albertkol) [reply templates/account/invoices/index.html]: We have the robots tag now.

The meta was updated too.

**Comment by albertkol** [comment]
> QA:
> 
> * I am getting a console error: `Uncaught TypeError: Currency code is required with currency style.` on the invoice page
> * The invoices themselves are 500'ing for me in the demo (I remember you saying something about that this morning)

Both were fixed. 500s were fixed with a temporary change:
- we show the Download PDF link text as "Download"
- we just redirect the user to the PDF location

As a separate step we will have custom download file name link text. It's just easier to have that as a separate PR.

**Comment by albertkol** [comment]
@ClementChaumel let me know if you do not like my front-end; if you envisioned to get the invoice data as an endpoint returning you a json for you to build everything with React. I leave it up to you.

I was "in the zone" last Wednesday, and ended up creating the front-end for the endpoint.

**Comment by ClementChaumel** [comment]
> @ClementChaumel let me know if you do not like my front-end; if you envisioned to get the invoice data as an endpoint returning you a json for you to build everything with React. I leave it up to you.
> 
> I was "in the zone" last Wednesday, and ended up creating the front-end for the endpoint.

I think it's perfectly fine to display this view using Jinja templates. LGTM :+1:

**Comment by elioqoshi** [comment]
Thanks @albertkol !

A few minor things:

- (This is also a visual thing) When there are a few invoices, the minimum height of the invoice section ends right at the last invoice. We should keep a minimal height which is larger than that, maybe something like this:

image

- Probably also something for @lechsandecki to review later on, though "Canonical UA" feels quite abbrevated. Any reason we are not going with "Ubuntu Advantage"? I might have missed something here.

**Comment by albertkol** [comment]
@elioqoshi Added a bit of spacing.

**Comment by elioqoshi** [comment]
@albertkol Thanks! Looks good to me!

Out of curiousity, is there any reason for this break point on mobile? 
image

Having a grid layout of 2 columns seems a bit weird for listing invoices and 1 vertical column should be the case on all resolutions. But I don't know if that's a Vanilla limitation. Not making this a blocker though

**Comment by elioqoshi** [comment]
So with the new API change, clicking on downloading invoice forwards the user to a stripe page where the invoice information is available and from where you can download the invoice:

image

I do have a few concerns with this:

- Navigating to another page as a response to `Download invoice` is not the right expected behaviour. It would add unnecessary complexity UNLESS the stripe invoice page has some additional information we want to surface (e.g what is the difference between downloading the invoice and receipt?).

I'd suggest to allow the user to download the invoice right away unless there is a good reason we are surfacing the stripe page.

**Comment by albertkol** [comment]
@elioqoshi the download link was reverted to downloading the invoice directly. I hope you do not mind if I revert the -1 to +1 back.

**Review by ClementChaumel — APPROVED** [review]

**Pull request merged**
